core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	for _, fwdr := range fwdrs {
+		EOAs, err := f.getContractSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnf("Failed to get forwarder senders %s", err)
+			continue
+		}
+		for _, EOA := range EOAs {
+			if EOA == from {
+				forwardedPayload, err := f.getForwardedPayload(to, encodedPayload)
+				if err != nil {
+					f.logger.AssumptionViolation("Forwarder encoding failed, this should never happen")
+					continue
+				}
+				return fwdr.Address, forwardedPayload, nil
+			}
+		}
+	}
+
+	return to, encodedPayload, errors.New("Skipping forwarding transaction")
+}
+
+func (f *FwdMgr) getForwardedPayload(dest common.Address, origPayload []byte) ([]byte, error) {
+	callArgs, err := forwardABI.Inputs.Pack(dest, origPayload)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to pack forwarder payload")
+	}
+
+	dataBytes := append(forwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to call getAuthorizedSenders on %s", addr)
+	}
+	f.setCachedSenders(addr, senders)
+	f.subscribeSendersChangedLogs(addr)
+	return senders, nil
+}
+
+func (f *FwdMgr) getAuthorizedSenders(addr common.Address) ([]common.Address, error) {
+	c, err := authorized_receiver.NewAuthorizedReceiverCaller(addr, f.evmClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to init forwarder caller")
+	}
+	opts := bind.CallOpts{Context: f.ctx, Pending: false}
+	senders, err := c.GetAuthorizedSenders(&opts)
+	if err != nil {
+		return nil, err
+	}
+	return senders, nil
+}
+
+func (f *FwdMgr) initForwardersCache(ctx context.Context, fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		senders, err := f.getAuthorizedSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnw("Failed to call getAuthorizedSenders on forwarder", fwdr, "err", err)
+			continue
+		}
+		f.setCachedSenders(fwdr.Address, senders)
+
+	}
+}
+func (f *FwdMgr) subscribeForwardersLogs(fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		f.subscribeSendersChangedLogs(fwdr.Address)
+	}
+}
+
+func (f *FwdMgr) subscribeSendersChangedLogs(addr common.Address) {
+	f.logpoller.MergeFilter(
+		[]common.Hash{authorized_receiver.AuthorizedReceiverAuthorizedSendersChanged{}.Topic()},
+		addr)
+}
+
+func (f *FwdMgr) setCachedSenders(addr common.Address, senders []common.Address) {
+	f.cacheMu.Lock()
+	defer f.cacheMu.Unlock()
+	f.sendersCache[addr] = senders
+}
+
+func (f *FwdMgr) getCachedSenders(addr common.Address) ([]common.Address, bool) {
+	f.cacheMu.RLock()
+	defer f.cacheMu.RUnlock()
+	addrs, ok := f.sendersCache[addr]
+	return addrs, ok
+}
+
+func (f *FwdMgr) runLoop() {
+	defer f.wg.Done()
+	tick := time.After(0)
+
+	for {
+		select {
+		case <-tick:
+			tick = time.After(utils.WithJitter(authLogCheckInterval))
```

Comment:
If the work below starts running very slowly (~`authLogCheckInterval`), would it be better to wait to start the next tick until afterwards, instead of potentially running immeadiately again?
```suggestion
	for ; ; tick = time.After(utils.WithJitter(authLogCheckInterval)) {
		select {
		case <-tick:
```

Reply:
Can't this just be done by setting the tick right after finishing everything in the case?

Reply:
I set the tick after the db query, the log check interval should be quite longer than a couple of seconds (I set it to 1min, but it should be longer than that. Will try to discuss with  steve on the best way to ration about this)

Reply:
I thought it should be after everything, including:
```go
for _, log := range logs {
	if err = f.handleAuthChange(log); err != nil {
```
and was just accounting for the continues - could refactor lots of different ways though :shrug:

Reply:
It's not a big change, done